refactor: adds concurrent plugin invocations in plugin-based actions

jpower432 opened the pull request:
This update allows plugin-based operations to be executed concurrently using a fan out, fan in approach.
This is benefical when scaling up the number of plugins used, espcially plugins with longer running scan times.

Closes #137

marcusburghardt commented on framework/actions/input.go:
```diff
@@ -35,12 +35,15 @@ type InputContext struct {
 	inputCtx := &InputContext{
-		requestedProviders: make(map[plugin.ID]string),
+		requestedProviders:  make(map[plugin.ID]string),
+		MaxConcurrentWeight: 3,
```
Could be make this limit configurable in runtime?

jpower432 replied:
Thanks for the feedback @marcusburghardt. To clarify, you mean as an option in the CLI or the config?

jpower432 replied:
Added this option and the timeout value as config options

marcusburghardt commented on framework/actions/aggregate.go:
```diff
@@ -21,24 +24,68 @@ import (
+
+		wg.Add(1)
+
+		go func(providerId plugin.ID, plugin policy.Provider) {
```
I'm wondering if it wouldn't be beneficial to include a timeout in this function in case a plugin stalls.

marcusburghardt replied:
In expected cases we can speed up the assessment, which is great. But I also wonder if we should abort the whole process instead of returning incomplete results. One advantage with this parallelism is that we can abort earlier and save resources unless partial results bring value.

marcusburghardt commented on framework/actions/aggregate_test.go:
```diff
@@ -89,6 +94,79 @@ func TestAggregateResults(t *testing.T) {
 	require.Len(t, gotResults, 1)
 }
 
+func TestAggregateResults_Multi(t *testing.T) {
```
A test scenario where one plugin fails could be good to see this behavior in our tests.

marcusburghardt reviewed (commented):
Nice work @jpower432 . Unit tests are not working in my local environment and I need to fix this on my side, but would like to already share some thoughts.

jpower432 commented:
@marcusburghardt PTAL. I switched to `errgroup` to simplify the solution and to better support context propagation to the goroutines. This change required a breaking change on the `policy.Provider` interface methods to accept a `context.Context` argument. I've also introduced a `pluginTimeout` with a value that I think is reasonable in the event a plugin need to also perform quick scans, but I'm open to feedback on that specific duration.

marcusburghardt reviewed (changes requested):
Nice @jpower432 . I have only a minor comment for the value used in the log message.

jpower432 commented on cmd/c2pcli/cli/subcommands/config.go:
```diff
@@ -54,6 +55,12 @@ func Context(ap *oscalTypes.AssessmentPlan) (*actions.InputContext, error) {
 
+	// Set the max concurrency if set by the user
+	if option.AdvancedOptions.MaxConcurrency != 0 {
+		option.logger.Debug("Setting max concurrency", "max", option.AdvancedOptions.MaxPluginTimeout)
```
Good catch! Thanks @marcusburghardt.

marcusburghardt reviewed (approved):
LGTM. Thanks @jpower432

jpower432 commented:
@gvauter @marcusburghardt Just sharing that the last push was to resolve a merge conflict. I will do a quick re-review before merging.

jpower432 commented:
Looks like there is a [flaky test](https://github.com/oscal-compass/compliance-to-policy-go/actions/runs/16521216452/attempts/1). I will take care of that first.
